File: ios/MitiMaiti/ViewModels/OnboardingViewModel.swift
import SwiftUI
import PhotosUI

@MainActor
class OnboardingViewModel: ObservableObject {
    @Published var currentStep: OnboardingStep = .name
    @Published var firstName = ""
    @Published var isNonSindhi = false
    @Published var birthDay = 15
    @Published var birthMonth = 6
    @Published var birthYear = 1998
    @Published var selectedGender: Gender?
    @Published var selectedPhotos: [String] = []
    @Published var selectedImages: [UIImage] = []
    @Published var selectedIntent: Intent?
    @Published var selectedShowMe: ShowMe?
    @Published var selectedCity: String?
    @Published var isLoading = false
    @Published var error: String?

    var age: Int {
        let components = DateComponents(year: birthYear, month: birthMonth, day: birthDay)
        guard let dob = Calendar.current.date(from: components) else { return 0 }
        return Calendar.current.dateComponents([.year], from: dob, to: Date()).year ?? 0
    }

    var isAgeValid: Bool {
        age >= 18
    }

    var progress: Double {
        Double(currentStep.rawValue + 1) / Double(OnboardingStep.allCases.count)
    }

    var canProceed: Bool {
        switch currentStep {
        case .name: return firstName.count >= 2
        case .birthday: return isAgeValid
        case .gender: return selectedGender != nil
        case .photos: return selectedImages.count >= 1
        case .intent: return selectedIntent != nil
        case .showMe: return selectedShowMe != nil
        case .location: return selectedCity != nil
        case .ready: return true
        }
    }

    func nextStep() {
        guard canProceed else { return }
        if let next = OnboardingStep(rawValue: currentStep.rawValue + 1) {
            withAnimation(.spring(response: 0.4, dampingFraction: 0.8)) {
                currentStep = next
            }
        }
    }

    func previousStep() {
        if let prev = OnboardingStep(rawValue: currentStep.rawValue - 1) {
            withAnimation(.spring(response: 0.4, dampingFraction: 0.8)) {
                currentStep = prev
            }
        }
    }

    func addPhoto() {
        if selectedPhotos.count < 6 {
            selectedPhotos.append("photo_\(selectedPhotos.count + 1)")
        }
    }

    func addImage(_ image: UIImage) {
        guard selectedImages.count < 6 else { return }
        selectedImages.append(image)
        selectedPhotos.append("photo_\(selectedPhotos.count + 1)")
        // Save the first photo as the profile picture
        if selectedImages.count == 1 {
            UserImageStore.shared.save(image)
        }
    }

    func removePhoto(at index: Int) {
        guard index < selectedPhotos.count else { return }
        selectedPhotos.remove(at: index)
        if index < selectedImages.count {
            selectedImages.remove(at: index)
        }
    }
}

File: android/app/src/main/java/com/mitimaiti/app/viewmodels/OnboardingViewModel.kt
package com.mitimaiti.app.viewmodels

import android.net.Uri
import androidx.lifecycle.ViewModel
import com.mitimaiti.app.models.Gender
import com.mitimaiti.app.models.Intent
import com.mitimaiti.app.models.ShowMe
import com.mitimaiti.app.services.PhotoRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

enum class OnboardingStep { NAME, BIRTHDAY, GENDER, PHOTOS, INTENT, SHOW_ME, LOCATION, READY }

class OnboardingViewModel : ViewModel() {
    private val _currentStep = MutableStateFlow(OnboardingStep.NAME)
    val currentStep: StateFlow<OnboardingStep> = _currentStep.asStateFlow()
    val firstName = MutableStateFlow(""); val isNonSindhi = MutableStateFlow(false)
    val birthDay = MutableStateFlow(""); val birthMonth = MutableStateFlow(""); val birthYear = MutableStateFlow("")
    val selectedGender = MutableStateFlow<Gender?>(null)
    val selectedPhotos: StateFlow<List<Uri>> = PhotoRepository.photos
    val selectedIntent = MutableStateFlow<Intent?>(null); val selectedShowMe = MutableStateFlow<ShowMe?>(null)
    val selectedCity = MutableStateFlow("")

    val age: Int? get() {
        val y = birthYear.value.toIntOrNull() ?: return null; val m = birthMonth.value.toIntOrNull() ?: return null; val d = birthDay.value.toIntOrNull() ?: return null
        return try { java.time.Period.between(java.time.LocalDate.of(y, m, d), java.time.LocalDate.now()).years } catch (e: Exception) { null }
    }
    val isAgeValid: Boolean get() = (age ?: 0) >= 18
    val progress: Float get() = (OnboardingStep.entries.indexOf(_currentStep.value) + 1).toFloat() / OnboardingStep.entries.size
    val canProceed: Boolean get() = when (_currentStep.value) {
        OnboardingStep.NAME -> firstName.value.isNotBlank(); OnboardingStep.BIRTHDAY -> isAgeValid
        OnboardingStep.GENDER -> selectedGender.value != null; OnboardingStep.PHOTOS -> selectedPhotos.value.isNotEmpty()
        OnboardingStep.INTENT -> selectedIntent.value != null; OnboardingStep.SHOW_ME -> selectedShowMe.value != null
        OnboardingStep.LOCATION -> selectedCity.value.isNotBlank(); OnboardingStep.READY -> true
    }

    fun nextStep() { val s = OnboardingStep.entries; val i = s.indexOf(_currentStep.value); if (i < s.size - 1) _currentStep.value = s[i + 1] }
    fun previousStep() { val s = OnboardingStep.entries; val i = s.indexOf(_currentStep.value); if (i > 0) _currentStep.value = s[i - 1] }
    fun addPhoto(uri: Uri) { PhotoRepository.addPhoto(uri) }
    fun removePhoto(index: Int) { PhotoRepository.removePhoto(index) }
}
